### DISCUSSION TITLE
Issue with aspire-starter on Linux

### DISCUSSION THREAD

**Discussion by sprudel79** [microsoft/aspire#4546] [Q&A] [answered]
Hi there, congrats first of all to the recent GA release of .NET Aspire!
I have been following the project for the last months and I'd like to give it a try on my local Linux environment.
The issue is this error when trying to run the aspire-starter project without any modifications:

![image](https://github.com/dotnet/aspire/assets/10754010/3cf48f0c-62b2-4ebc-ba7f-efd0bdd2d0fb)

In order to get more information about this error I have set:

`"Aspire.Hosting.Dcp": "Debug"`

in my application.json of the AppHost project which gives me the following logs:

```
dbug: Aspire.Hosting.Dcp.ApplicationExecutor[0]
      Starting watch over DCP Service resources
Aspire.Hosting.Dcp.ApplicationExecutor: Debug: Starting watch over DCP Container resources
dbug: Aspire.Hosting.Dcp.ApplicationExecutor[0]
      Starting watch over DCP Endpoint resources
dbug: Aspire.Hosting.Dcp.ApplicationExecutor[0]
      Starting watch over DCP Container resources
dbug: Aspire.Hosting.Dcp.ApplicationExecutor[0]
      Starting watch over DCP Executable resources
Aspire.Hosting.Dcp.ApplicationExecutor: Debug: Starting watch over DCP Service resources
Aspire.Hosting.Dcp.ApplicationExecutor: Debug: Starting watch over DCP Endpoint resources
Aspire.Hosting.Dcp.ApplicationExecutor: Debug: Starting watch over DCP Executable resources
dbug: Aspire.Hosting.Dcp.KubernetesService[0]
      Waiting for Kubernetes configuration file at '/tmp/aspire.04J4cA/kubeconfig' (attempt 0).
Aspire.Hosting.Dcp.KubernetesService: Debug: Waiting for Kubernetes configuration file at '/tmp/aspire.04J4cA/kubeconfig' (attempt 0).
dbug: Aspire.Hosting.Dcp.KubernetesService[0]
      Waiting for Kubernetes configuration file at '/tmp/aspire.04J4cA/kubeconfig' (attempt 1).
Aspire.Hosting.Dcp.KubernetesService: Debug: Waiting for Kubernetes configuration file at '/tmp/aspire.04J4cA/kubeconfig' (attempt 1).
Aspire.Hosting.Dcp.KubernetesService: Debug: Waiting for Kubernetes configuration file at '/tmp/aspire.04J4cA/kubeconfig' (attempt 2).
dbug: Aspire.Hosting.Dcp.KubernetesService[0]
Aspire.Hosting.Dcp.KubernetesService: Debug: Waiting for Kubernetes configuration file at '/tmp/aspire.04J4cA/kubeconfig' (attempt 7).
dbug: Aspire.Hosting.Dcp.KubernetesService[0]
      Waiting for Kubernetes configuration file at '/tmp/aspire.04J4cA/kubeconfig' (attempt 7).
Aspire.Hosting.Dcp.KubernetesService: Debug: Waiting for Kubernetes configuration file at '/tmp/aspire.04J4cA/kubeconfig' (attempt 8).
dbug: Aspire.Hosting.Dcp.KubernetesService[0]
      Waiting for Kubernetes configuration file at '/tmp/aspire.04J4cA/kubeconfig' (attempt 8).
info: Aspire.Hosting.Dcp.start-apiserver.api-server[0]
      Starting API server...
Aspire.Hosting.Dcp.start-apiserver.api-server: Information: Starting API server...
info: Aspire.Hosting.Dcp.start-apiserver.api-server[0]
      API server started	{"Address": "127.0.0.1", "Port": 37599}
Aspire.Hosting.Dcp.start-apiserver.api-server: Information: API server started	{"Address": "127.0.0.1", "Port": 37599}
Aspire.Hosting.Dcp.start-apiserver.dcp-host: Information: Starting DCP controller host
info: Aspire.Hosting.Dcp.start-apiserver.dcp-host[0]
      Starting DCP controller host
Aspire.Hosting.Dcp.start-apiserver.dcp-host: Information: Started all services	{"count": 1}
info: Aspire.Hosting.Dcp.start-apiserver.dcp-host[0]
      Started all services	{"count": 1}
info: Aspire.Hosting.Dcp.dcpctrl.IdeExecutableRunner[0]
      Executables cannot be started via IDE: missing required environment variable 'DEBUG_SESSION_PORT'
Aspire.Hosting.Dcp.dcpctrl.IdeExecutableRunner: Information: Executables cannot be started via IDE: missing required environment variable 'DEBUG_SESSION_PORT'
info: Aspire.Hosting.Dcp.dcpctrl[0]
      starting controller manager
Aspire.Hosting.Dcp.dcpctrl: Information: starting controller manager
```
So it looks like there's an issue with DCP that is required by Aspire (documented at https://github.com/dotnet/aspire/blob/main/docs/specs/IDE-execution.md). I assume there's no sense setting the the environment variable by myself.

Here are first of all the required details about my environment:

```
lsb_release -a
No LSB modules are available.
Distributor ID: Linuxmint
Description:    Linux Mint 21.3
Release:        21.3
Codename:       virginia

dotnet --list-sdks
8.0.301 [/usr/share/dotnet/sdk]

code --version
1.90.0
89de5a8d4d6205e5b11647eb6a74844ca23d2573
x64

docker version
Client: Docker Engine - Community
 Version:           26.1.4
 API version:       1.45
 Go version:        go1.21.11
 Git commit:        5650f9b
 Built:             Wed Jun  5 11:28:57 2024
 OS/Arch:           linux/amd64
 Context:           default

Server: Docker Engine - Community
 Engine:
  Version:          26.1.4
  API version:      1.45 (minimum version 1.24)
  Go version:       go1.21.11
  Git commit:       de5c9cf
  Built:            Wed Jun  5 11:28:57 2024
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.33
  GitCommit:        d2d58213f83a351ca8f528a95fbd145f5654e957
 runc:
  Version:          1.1.12
  GitCommit:        v1.1.12-0-g51d5e94
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

Next, I have installed .NET Aspire workload via:

```
sudo dotnet workload install aspire

dotnet workload list

Installed Workload Id      Manifest Version      Installation Source
--------------------
aspire                     8.0.1/8.0.100         SDK 8.0.300        

Use `dotnet workload search` to find additional workloads to install.
```

And finally, here are the exact steps that I did:

1. dotnet new aspire-starter -o aspiretest
2. cd aspiretest
3. code .
4. F5 
5. VSCode asks about the project to be launched and I have chosen "aspiretest.AppHost"
6. The solution builds and starts successfully and I will see the localhost link incl. token parameter to be clicked on which will open the browser tab as shown above in the screenshot

Was there anything wrong from my side? 
In general on https://learn.microsoft.com/en-us/dotnet/aspire/get-started/build-your-first-aspire-app?pivots=visual-studio I can read about the prerequisites of Aspire which doesn't tell about the operating system, also Visual Studio and even the C# Dev Kit extension for VSCode are listed as optional prerequisites.

Many thanks in advance for any kind of hint.

Just before sending the issue I have recognized the following section in the Aspire documentation (my first link above):

For IDE execution to work, two conditions need to be fulfilled:

- DCP needs to be told how to contact the IDE (what is the IDE session endpoint, specifically).
- The ExecutionType property for the Executable object needs to be set to IDE (default is Process, which indicates OS process-based execution).

How am I supposed to do that in VSCode?
Thanks again!

**Comment by davidfowl** [comment]
You don't need an IDE, that's for the enhanced experience. The command line should work just fine without one (VS or VS code). Can you see the error in the dashboard?

**Comment by sprudel79** [comment, marked as answer]
Hi @davidfowl , thanks for your instant reply, I was able to solve it.
I have tried it again first of all via CLI only and have followed exactly the steps described here:

https://learn.microsoft.com/en-us/dotnet/aspire/get-started/build-your-first-aspire-app?pivots=dotnet-cli

The most important thing for me was that line:
**If you haven't already trusted the ASP.NET Core localhost certificate, you will need to trust the certificate before running the app:**

Since I am a Linux user, I have followed a few links and landed right here:

https://learn.microsoft.com/en-us/aspnet/core/security/enforcing-ssl?view=aspnetcore-8.0&tabs=visual-studio%2Clinux-ubuntu#trust-https-certificate-on-linux

I have executed the following commands:

```
dotnet tool update -g linux-dev-certs

You can invoke the tool using the following command: dotnet-linux-dev-certs
Tool 'linux-dev-certs' (version '0.2.0') was successfully installed.

dotnet linux-dev-certs install

Some operations require root. You may be prompted for your 'sudo' password.
Creating CA certificate.
Installing CA certificate.
[sudo] password for user:           
Removing existing development certificates.
Creating development certificate.
Installing development certificate.
The development certificate was successfully installed.

```

I still get a warning for in my browser that the ASP.NET Core developer certificate is not trusted but there's no "unhandled error" anymore on the dashboard and it properly loads it now:

![image](https://github.com/dotnet/aspire/assets/10754010/8e0cd5cc-63f0-4725-b697-1d88b4b3cc33)

So I am very happy that it's working now on my side.
I assume it's solved due to the CA being installed, but perhaps you could please comment about the requirements of .NET Aspire regarding certificates / https, many thanks in advance!

**Reply by davidfowl** [reply]
> but perhaps you could please comment about the requirements of .NET Aspire regarding certificates / https, many thanks in advance!

Are you talking about all of the moving pieces and how they use HTTPS?

**Reply by sprudel79** [reply]
Exactly, I am wondering how .NET Aspire deals with HTTPS and if all the moving pieces require the use of HTTPS to work correctly (which, in my opinion, would be perfectly fine) or if a configuration with HTTP is also possible (even it might be pointless). Thanks.

**Reply by davidfowl** [reply]
The documentation here is very light but there are several user facing components that are using HTTPS and the ASP.NET Core developer cert:

https://learn.microsoft.com/en-us/dotnet/aspire/fundamentals/dashboard/security-considerations
https://learn.microsoft.com/en-us/dotnet/aspire/troubleshooting/allow-unsecure-transport

All of the components that aspire ships use HTTPS by default. The developer certificate is used by the dashboard and the apphost project (which the dashboard connects to in order to show information).

**Reply by sprudel79** [reply]
Hi @davidfowl , thank you, exactly what I was looking for.
To summarize, I will ensure that I have a proper HTTPS/certificate setup on my local machine (although it is possible to run it in an insecure manner). On Linux, trusting the developer certificate properly is a bit more tedious, but the process is well described in the second link I mentioned above.

**Reply by davidfowl** [reply]
> On Linux, trusting the developer certificate properly is a bit more tedious, but the process is well described in the second link I mentioned above.

We're working to improve this in .NET 9.
